fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerConflictsTest.java

```diff
@@ -0,0 +1,205 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.provider.foundationdb;
+
+import com.apple.foundationdb.record.IndexState;
+import com.apple.foundationdb.record.TestRecords1Proto;
+import com.apple.foundationdb.record.metadata.Index;
+import com.apple.foundationdb.record.metadata.IndexTypes;
+import com.apple.test.Tags;
+import org.junit.jupiter.api.Tag;
+import org.junit.jupiter.api.Test;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+import java.util.Random;
+import java.util.stream.Collectors;
+import java.util.stream.LongStream;
+
+import static com.apple.foundationdb.record.metadata.Key.Expressions.field;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertSame;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+/**
+ * Tests for {@link OnlineIndexer}. Checking different db manipulations the indexing process.
+ */
+
+@Tag(Tags.RequiresFDB)
+public class OnlineIndexerConflictsTest extends OnlineIndexerTest {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(OnlineIndexerConflictsTest.class);
+
+    @Test
+    public void testOnlineIndexerIdempotent() throws Exception {
+
+        Random rand = new Random();
+        final long seed = rand.nextInt();
+        LOGGER.info("Random seed is " + seed);
+        rand = new Random(seed); // make rand repeatable (note that multi threading behaviour is not fully repeatable in this test)
+
+        List<TestRecords1Proto.MySimpleRecord> records =
+                LongStream.range(0, 400).mapToObj(val -> TestRecords1Proto.MySimpleRecord.newBuilder().setRecNo(val).setNumValue2((int)val + 1).build()
+                                                 ).collect(Collectors.toList());
+
+        Index index = new Index("newIndex", field("num_value_2"), IndexTypes.VALUE);
+        FDBRecordStoreTestBase.RecordMetaDataHook hook = metaDataBuilder -> metaDataBuilder.addIndex("MySimpleRecord", index);
+        openSimpleMetaData();
+        try (FDBRecordContext context = openContext()) {
+            for (int i = 50; i <= 150; i++) {
+                // even numbers 100-300 (that's 101 values)
+                recordStore.saveRecord(records.get(i * 2));
+            }
+            context.commit();
+        }
+
+        openSimpleMetaData(hook);
+        try (FDBRecordContext context = openContext()) {
+            assertTrue(recordStore.uncheckedMarkIndexReadable(index.getName()).join());
```

Comment:
Hmm. IIUC from offline conversations, this was needed to make future commits down the line not fail with a conflict, but this call seems suspicious to me. It essentially marks the index as already built, so the index build won't do much of anything, so it seems more likely to me that this is either masking some other problem, or there is some issue or is a sign that there is some other kind of misconfiguration going on with these tests.

Reply:
I agree. This line was copied from the tests in OnlineIndexerSimpleTest. Would you have any idea why the indexer commit was failing without this line?

Reply:
I believe all of the instances of that method being used in `OnlineIndexerSimpleTest`, it does so so that it can then scan the index without either (1) being told that the index isn't readable (which is what happens if you scan an unbuilt index normally) or (2) building the index and therefore disturbing it. It also looks like in all of the instances (that I could find) in `OnlinIndexerSimpleTest`, the transaction that calls `uncheckedMarkIndexReadable` isn't committed (and indeed, there are comments saying `// do NOT commit` instead of calls to `commit()`).

I'm not sure why it's currently failing without that change, though. I had some theories, but most of the theories ended up being falsified by further experiments that you did. It's possible this needs more investigation, which might mean turning on detailed FDB client trace logging, or possibly investing in #1002 (or something temporary along those lines) to see what the conflict is.

Reply:
I've added the test you've suggested above. Here I could mark the index as write only.